### PR TITLE
compile_and_test_for_board.py: add optional JUnit XML support

### LINKED ISSUE #14620: compile_and_test_for_board.py: provide xunit file?
Does it make sense to add optional xunit XML support for `compile_and_test_for_board.py`? This way the results of #14587 could be folded into the results of https://github.com/RIOT-OS/Release-Specs/pull/174 and the nightly HiL tests.

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
This implements my proposal from #14620.

### Testing procedure
Install [`junit-xml`](https://pypi.org/project/junit-xml/) and run `./compile_and_test_for_board.py` for a board of your choice using the `--report-xml` argument (and without also to check if it does not break). When the `--report-xml` argument is provided, there should be a `report.xml` in the result directory that lists the results that are also present in the directory in correct jUnit XML. When `--report-xml` is not provided, `report.xml` is not generated (but might still be present from a previous run).

### Issues/PRs references
Resolves #14620.

### CODE REVIEW

**Comment by MrKevinWeiss** [comment]
I will try it in the morning.

**Comment by MrKevinWeiss** [comment]
Currently I am getting an error trying to use it

<details>

```
./dist/tools/compile_and_test_for_board/compile_and_test_for_board.py . samr21-xpro --applications "tests/xtimer_periodic_wakeup tests/bloom_bytes" --report-xml
INFO:samr21-xpro:Saving toolchain
INFO:samr21-xpro.tests/bloom_bytes:Board supported: True
INFO:samr21-xpro.tests/bloom_bytes:Board has enough memory: True
INFO:samr21-xpro.tests/bloom_bytes:Application has test: True
INFO:samr21-xpro.tests/bloom_bytes:Run compilation
INFO:samr21-xpro.tests/bloom_bytes:Run test
INFO:samr21-xpro.tests/bloom_bytes:Run test.flash
INFO:samr21-xpro.tests/bloom_bytes:Success
INFO:samr21-xpro.tests/xtimer_periodic_wakeup:Board supported: True
INFO:samr21-xpro.tests/xtimer_periodic_wakeup:Board has enough memory: True
INFO:samr21-xpro.tests/xtimer_periodic_wakeup:Application has test: True
INFO:samr21-xpro.tests/xtimer_periodic_wakeup:Run compilation
INFO:samr21-xpro.tests/xtimer_periodic_wakeup:Run test
INFO:samr21-xpro.tests/xtimer_periodic_wakeup:Run test.flash
INFO:samr21-xpro.tests/xtimer_periodic_wakeup:Success
Traceback (most recent call last):
  File "./dist/tools/compile_and_test_for_board/compile_and_test_for_board.py", line 768, in <module>
    main(PARSER.parse_args())
  File "./dist/tools/compile_and_test_for_board/compile_and_test_for_board.py", line 753, in main
    junit_xml.TestSuite.to_file(
  File "/home/weiss/.local/lib/python3.8/site-packages/junit_xml/__init__.py", line 272, in to_file
    to_xml_report_file(file_descriptor, test_suites, prettyprint, encoding)
  File "/home/weiss/.local/lib/python3.8/site-packages/junit_xml/__init__.py", line 320, in to_xml_report_file
    xml_string = to_xml_report_string(test_suites, prettyprint=prettyprint, encoding=encoding)
  File "/home/weiss/.local/lib/python3.8/site-packages/junit_xml/__init__.py", line 290, in to_xml_report_string
    ts_xml = ts.build_xml_doc(encoding=encoding)
  File "/home/weiss/.local/lib/python3.8/site-packages/junit_xml/__init__.py", line 127, in build_xml_doc
    if any(c.assertions for c in self.test_cases):
  File "/home/weiss/.local/lib/python3.8/site-packages/junit_xml/__init__.py", line 127, in <genexpr>
    if any(c.assertions for c in self.test_cases):
AttributeError: 'str' object has no attribute 'assertions
```
</details>

I have not ruled out the fact that I may be human so let just double check it isn't something silly I have done.

**Comment by MrKevinWeiss** [comment]
hmmm both python 3.7 and 3.7 are the same.  Does

`./dist/tools/compile_and_test_for_board/compile_and_test_for_board.py . samr21-xpro --applications "tests/xtimer_periodic_wakeup" --report-xml` work for you?

**Comment by miri64** [comment]
Mhhh... no. Will investigate.

**Comment by miri64** [comment]
Fixed. Not really sure why, but there was a stray argument, that did not really belong their.

**Comment by miri64** [comment]
This is what the `results/samr21-xpro/report.xml` should look like for the output you provided btw.

<details>

```xml
<?xml version="1.0" ?>
<testsuites disabled="0" errors="0" failures="0" tests="1" time="9.420432329177856">
	<testsuite disabled="0" errors="0" failures="0" name="compile_and_test_for_samr21-xpro" skipped="0" tests="1" time="9.420432329177856">
		<testcase name="tests/xtimer_periodic_wakeup" time="9.420432" timestamp="1598461777.8198678" log="INFO:samr21-xpro.tests/xtimer_periodic_wakeup:Board supported: True
INFO:samr21-xpro.tests/xtimer_periodic_wakeup:Board has enough memory: True
INFO:samr21-xpro.tests/xtimer_periodic_wakeup:Application has test: True
INFO:samr21-xpro.tests/xtimer_periodic_wakeup:Run compilation
INFO:samr21-xpro.tests/xtimer_periodic_wakeup:Run test
INFO:samr21-xpro.tests/xtimer_periodic_wakeup:Run test.flash
INFO:samr21-xpro.tests/xtimer_periodic_wakeup:Success
">
			<system-out>
r
/home/mlenders/Repositories/RIOT-OS/RIOT/dist/tools/pyterm/pyterm -p &quot;/dev/ttyACM0&quot; -b &quot;115200&quot; --no-reconnect --noprefix --no-repeat-command-on-empty-line 
Twisted not available, please install it if you want to use pyterm's JSON capabilities
Connect to serial port /dev/ttyACM0
Welcome to pyterm!
Type '/exit' to exit.
READY
s
START
main(): This is RIOT! (Version: buildtest)
xtimer_periodic_wakeup test application.

Testing interval 256... (now=3115419)
Testing interval 255... (now=3119184)
Testing interval 254... (now=3122948)
Testing interval 253... (now=3126711)
Testing interval 252... (now=3130473)
Testing interval 251... (now=3134234)
Testing interval 250... (now=3137994)
Testing interval 249... (now=3141753)
Testing interval 248... (now=3145511)
Testing interval 247... (now=3149268)
Testing interval 246... (now=3153024)
Testing interval 245... (now=3156779)
Testing interval 244... (now=3160533)
Testing interval 243... (now=3164285)
Testing interval 242... (now=3168037)
Testing interval 241... (now=3171787)
Testing interval 240... (now=3175537)
Testing interval 239... (now=3179285)
Testing interval 238... (now=3183033)
Testing interval 237... (now=3186780)
Testing interval 236... (now=3190525)
Testing interval 235... (now=3194270)
Testing interval 234... (now=3198014)
Testing interval 233... (now=3201757)
Testing interval 232... (now=3205497)
Testing interval 231... (now=3209237)
Testing interval 230... (now=3212976)
Testing interval 229... (now=3216715)
Testing interval 228... (now=3220453)
Testing interval 227... (now=3224189)
Testing interval 226... (now=3227925)
Testing interval 225... (now=3231660)
Testing interval 224... (now=3235393)
Testing interval 223... (now=3239126)
Testing interval 222... (now=3242858)
Testing interval 221... (now=3246589)
Testing interval 220... (now=3250319)
Testing interval 219... (now=3254047)
Testing interval 218... (now=3257774)
Testing interval 217... (now=3261500)
Testing interval 216... (now=3265225)
Testing interval 215... (now=3268950)
Testing interval 214... (now=3272674)
Testing interval 213... (now=3276397)
Testing interval 212... (now=3280119)
Testing interval 211... (now=3283839)
Testing interval 210... (now=3287559)
Testing interval 209... (now=3291277)
Testing interval 208... (now=3294994)
Testing interval 207... (now=3298711)
Testing interval 206... (now=3302426)
Testing interval 205... (now=3306140)
Testing interval 204... (now=3309852)
Testing interval 203... (now=3313564)
Testing interval 202... (now=3317275)
Testing interval 201... (now=3320985)
Testing interval 200... (now=3324693)
Testing interval 199... (now=3328401)
Testing interval 198... (now=3332108)
Testing interval 197... (now=3335815)
Testing interval 196... (now=3339521)
Testing interval 195... (now=3343225)
Testing interval 194... (now=3346929)
Testing interval 193... (now=3350632)
Testing interval 192... (now=3354333)
Testing interval 191... (now=3358034)
Testing interval 190... (now=3361733)
Testing interval 189... (now=3365431)
Testing interval 188... (now=3369129)
Testing interval 187... (now=3372825)
Testing interval 186... (now=3376521)
Testing interval 185... (now=3380215)
Testing interval 184... (now=3383908)
Testing interval 183... (now=3387601)
Testing interval 182... (now=3391292)
Testing interval 181... (now=3394983)
Testing interval 180... (now=3398673)
Testing interval 179... (now=3402361)
Testing interval 178... (now=3406048)
Testing interval 177... (now=3409734)
Testing interval 176... (now=3413419)
Testing interval 175... (now=3417104)
Testing interval 174... (now=3420788)
Testing interval 173... (now=3424471)
Testing interval 172... (now=3428153)
Testing interval 171... (now=3431834)
Testing interval 170... (now=3435514)
Testing interval 169... (now=3439193)
Testing interval 168... (now=3442871)
Testing interval 167... (now=3446548)
Testing interval 166... (now=3450224)
Testing interval 165... (now=3453898)
Testing interval 164... (now=3457572)
Testing interval 163... (now=3461245)
Testing interval 162... (now=3464917)
Testing interval 161... (now=3468588)
Testing interval 160... (now=3472258)
Testing interval 159... (now=3475925)
Testing interval 158... (now=3479592)
Testing interval 157... (now=3483258)
Testing interval 156... (now=3486923)
Testing interval 155... (now=3490587)
Testing interval 154... (now=3494250)
Testing interval 153... (now=3497913)
Testing interval 152... (now=3501574)
Testing interval 151... (now=3505234)
Testing interval 150... (now=3508893)
Testing interval 149... (now=3512551)
Testing interval 148... (now=3516208)
Testing interval 147... (now=3519863)
Testing interval 146... (now=3523519)
Testing interval 145... (now=3527173)
Testing interval 144... (now=3530826)
Testing interval 143... (now=3534478)
Testing interval 142... (now=3538129)
Testing interval 141... (now=3541779)
Testing interval 140... (now=3545428)
Testing interval 139... (now=3549076)
Testing interval 138... (now=3552723)
Testing interval 137... (now=3556369)
Testing interval 136... (now=3560014)
Testing interval 135... (now=3563657)
Testing interval 134... (now=3567300)
Testing interval 133... (now=3570942)
Testing interval 132... (now=3574583)
Testing interval 131... (now=3578223)
Testing interval 130... (now=3581862)
Testing interval 129... (now=3585500)
Testing interval 128... (now=3589136)
Testing interval 127... (now=3592772)
Testing interval 126... (now=3596407)
Testing interval 125... (now=3600041)
Testing interval 124... (now=3603672)
Testing interval 123... (now=3607303)
Testing interval 122... (now=3610933)
Testing interval 121... (now=3614563)
Testing interval 120... (now=3618192)
Testing interval 119... (now=3621819)
Testing interval 118... (now=3625446)
Testing interval 117... (now=3629072)
Testing interval 116... (now=3632697)
Testing interval 115... (now=3636321)
Testing interval 114... (now=3639944)
Testing interval 113... (now=3643567)
Testing interval 112... (now=3647188)
Testing interval 111... (now=3650808)
Testing interval 110... (now=3654426)
Testing interval 109... (now=3658044)
Testing interval 108... (now=3661660)
Testing interval 107... (now=3665275)
Testing interval 106... (now=3668890)
Testing interval 105... (now=3672503)
Testing interval 104... (now=3676115)
Testing interval 103... (now=3679727)
Testing interval 102... (now=3683338)
Testing interval 101... (now=3686948)
Testing interval 100... (now=3690556)
Testing interval 99... (now=3694163)
Testing interval 98... (now=3697679)
Testing interval 97... (now=3701194)
Testing interval 96... (now=3704708)
Testing interval 95... (now=3708222)
Testing interval 94... (now=3711734)
Testing interval 93... (now=3715246)
Testing interval 92... (now=3718756)
Testing interval 91... (now=3722265)
Testing interval 90... (now=3725773)
Testing interval 89... (now=3729280)
Testing interval 88... (now=3732786)
Testing interval 87... (now=3736291)
Testing interval 86... (now=3739796)
Testing interval 85... (now=3743300)
Testing interval 84... (now=3746802)
Testing interval 83... (now=3750303)
Testing interval 82... (now=3753803)
Testing interval 81... (now=3757302)
Testing interval 80... (now=3760800)
Testing interval 79... (now=3764295)
Testing interval 78... (now=3767792)
Testing interval 77... (now=3771288)
Testing interval 76... (now=3774783)
Testing interval 75... (now=3778277)
Testing interval 74... (now=3781770)
Testing interval 73... (now=3785261)
Testing interval 72... (now=3788752)
Testing interval 71... (now=3792242)
Testing interval 70... (now=3795730)
Testing interval 69... (now=3799218)
Testing interval 68... (now=3802705)
Testing interval 67... (now=3806190)
Testing interval 66... (now=3809674)
Testing interval 65... (now=3813157)
Testing interval 64... (now=3816640)
Testing interval 63... (now=3820121)
Testing interval 62... (now=3823601)
Testing interval 61... (now=3827080)
Testing interval 60... (now=3830558)
Testing interval 59... (now=3834035)
Testing interval 58... (now=3837511)
Testing interval 57... (now=3840987)
Testing interval 56... (now=3844460)
Testing interval 55... (now=3847932)
Testing interval 54... (now=3851405)
Testing interval 53... (now=3854876)
Testing interval 52... (now=3858345)
Testing interval 51... (now=3861815)
Testing interval 50... (now=3865283)
Testing interval 49... (now=3868750)
Testing interval 48... (now=3872217)
Testing interval 47... (now=3875682)
Testing interval 46... (now=3879146)
Testing interval 45... (now=3882610)
Testing interval 44... (now=3886072)
Testing interval 43... (now=3889533)
Testing interval 42... (now=3892994)
Testing interval 41... (now=3896454)
Testing interval 40... (now=3899913)
Testing interval 39... (now=3903370)
Testing interval 38... (now=3906827)
Testing interval 37... (now=3910244)
Testing interval 36... (now=3913654)
Testing interval 35... (now=3917063)
Testing interval 34... (now=3920473)
Testing interval 33... (now=3923882)
Testing interval 32... (now=3927292)
Testing interval 31... (now=3930701)
Testing interval 30... (now=3934111)
Testing interval 29... (now=3937513)
Testing interval 28... (now=3940916)
Testing interval 27... (now=3944318)
Testing interval 26... (now=3947721)
Testing interval 25... (now=3951123)
Testing interval 24... (now=3954526)
Testing interval 23... (now=3957929)
Testing interval 22... (now=3961325)
Testing interval 21... (now=3964721)
Testing interval 20... (now=3968117)
Testing interval 19... (now=3971512)
Testing interval 18... (now=3974908)
Testing interval 17... (now=3978304)
Testing interval 16... (now=3981699)
Testing interval 15... (now=3985088)
Testing interval 14... (now=3988477)
Testing interval 13... (now=3991866)
Testing interval 12... (now=3995255)
Testing interval 11... (now=3998643)
Testing interval 10... (now=4002032)
Testing interval 9... (now=4005418)
Testing interval 8... (now=4008695)
Testing interval 7... (now=4011972)
Testing interval 6... (now=4015249)
Testing interval 5... (now=4018526)
Testing interval 4... (now=4021803)
Testing interval 3... (now=4025080)
Testing interval 2... (now=4028357)
Testing interval 1... (now=4031635)
 256 diff=67
 255 diff=67
 254 diff=67
 253 diff=67
 252 diff=67
 251 diff=67
 250 diff=67
 249 diff=67
 248 diff=67
 247 diff=67
 246 diff=67
 245 diff=67
 244 diff=67
 243 diff=67
 242 diff=67
 241 diff=67
 240 diff=67
 239 diff=67
 238 diff=67
 237 diff=67
 236 diff=67
 235 diff=67
 234 diff=67
 233 diff=67
 232 diff=67
 231 diff=67
 230 diff=67
 229 diff=67
 228 diff=67
 227 diff=67
 226 diff=67
 225 diff=67
 224 diff=67
 223 diff=67
 222 diff=67
 221 diff=67
 220 diff=67
 219 diff=67
 218 diff=67
 217 diff=67
 216 diff=67
 215 diff=67
 214 diff=67
 213 diff=67
 212 diff=67
 211 diff=67
 210 diff=67
 209 diff=67
 208 diff=67
 207 diff=67
 206 diff=67
 205 diff=67
 204 diff=67
 203 diff=67
 202 diff=67
 201 diff=67
 200 diff=67
 199 diff=67
 198 diff=67
 197 diff=67
 196 diff=67
 195 diff=67
 194 diff=67
 193 diff=67
 192 diff=67
 191 diff=67
 190 diff=67
 189 diff=67
 188 diff=67
 187 diff=67
 186 diff=67
 185 diff=67
 184 diff=67
 183 diff=67
 182 diff=67
 181 diff=67
 180 diff=67
 179 diff=67
 178 diff=67
 177 diff=67
 176 diff=67
 175 diff=67
 174 diff=67
 173 diff=67
 172 diff=67
 171 diff=67
 170 diff=67
 169 diff=67
 168 diff=67
 167 diff=67
 166 diff=67
 165 diff=67
 164 diff=67
 163 diff=67
 162 diff=67
 161 diff=67
 160 diff=67
 159 diff=67
 158 diff=67
 157 diff=67
 156 diff=67
 155 diff=67
 154 diff=67
 153 diff=67
 152 diff=67
 151 diff=67
 150 diff=67
 149 diff=67
 148 diff=67
 147 diff=67
 146 diff=67
 145 diff=67
 144 diff=67
 143 diff=67
 142 diff=67
 141 diff=67
 140 diff=67
 139 diff=67
 138 diff=67
 137 diff=67
 136 diff=67
 135 diff=67
 134 diff=67
 133 diff=67
 132 diff=67
 131 diff=67
 130 diff=67
 129 diff=67
 128 diff=67
 127 diff=67
 126 diff=67
 125 diff=67
 124 diff=67
 123 diff=67
 122 diff=67
 121 diff=67
 120 diff=67
 119 diff=67
 118 diff=67
 117 diff=67
 116 diff=67
 115 diff=67
 114 diff=67
 113 diff=67
 112 diff=67
 111 diff=67
 110 diff=67
 109 diff=67
 108 diff=67
 107 diff=67
 106 diff=67
 105 diff=67
 104 diff=67
 103 diff=67
 102 diff=67
 101 diff=67
 100 diff=67
  99 diff=67
  98 diff=67
  97 diff=67
  96 diff=67
  95 diff=67
  94 diff=67
  93 diff=67
  92 diff=67
  91 diff=67
  90 diff=67
  89 diff=67
  88 diff=67
  87 diff=67
  86 diff=67
  85 diff=67
  84 diff=67
  83 diff=67
  82 diff=67
  81 diff=67
  80 diff=67
  79 diff=67
  78 diff=67
  77 diff=67
  76 diff=67
  75 diff=67
  74 diff=67
  73 diff=67
  72 diff=67
  71 diff=67
  70 diff=67
  69 diff=67
  68 diff=67
  67 diff=67
  66 diff=67
  65 diff=67
  64 diff=67
  63 diff=67
  62 diff=67
  61 diff=67
  60 diff=67
  59 diff=67
  58 diff=67
  57 diff=67
  56 diff=66
  55 diff=67
  54 diff=67
  53 diff=66
  52 diff=67
  51 diff=66
  50 diff=67
  49 diff=67
  48 diff=67
  47 diff=67
  46 diff=67
  45 diff=67
  44 diff=67
  43 diff=67
  42 diff=67
  41 diff=67
  40 diff=67
  39 diff=67
  38 diff=28
  37 diff=22
  36 diff=23
  35 diff=24
  34 diff=25
  33 diff=26
  32 diff=27
  31 diff=28
  30 diff=22
  29 diff=23
  28 diff=24
  27 diff=25
  26 diff=26
  25 diff=27
  24 diff=28
  23 diff=22
  22 diff=23
  21 diff=24
  20 diff=25
  19 diff=26
  18 diff=27
  17 diff=28
  16 diff=22
  15 diff=23
  14 diff=24
  13 diff=25
  12 diff=26
  11 diff=27
  10 diff=28
   9 diff=9
   8 diff=10
   7 diff=11
   6 diff=12
   5 diff=13
   4 diff=14
   3 diff=15
   2 diff=16
   1 diff=17

Min/max error: 9/67

Test complete.


</system-out>
		</testcase>
	</testsuite>
</testsuites>
```

</details>

File: dist/tools/compile_and_test_for_board/compile_and_test_for_board.py [hunk 1]
```diff
@@ -303,11 +328,29 @@ def run_compilation_and_test(self, **test_kwargs):
...
         try:
+            if self.testcase:
+                self.testcase.timestamp = time.time()
             self.compilation_and_test(**test_kwargs)
-            return None
+            res = None
         except ErrorInTest as err:
             self.logger.error('Failed during: %s', err)
-            return (str(err), err.application.appdir, err.errorfile)
+            res = (str(err), err.application.appdir, err.errorfile)
```
Comment (MrKevinWeiss) [inline dist/tools/compile_and_test_for_board/compile_and_test_for_board.py:337]: I think we should distinguish between `pass failure error skipped`. It should be an error if flashing or building fails but if the test fails I think it should be a `failure`.

**Review by MrKevinWeiss — CHANGES_REQUESTED** [review]
One comment, but generally good, tested pass case, failed test case, and skipped test case.

**Comment by MrKevinWeiss** [comment]
[example of skipping, passing and failing](https://riot-ci.inet.haw-hamburg.de/hil/job/RIOT-HIL/job/nightly/7/artifact/build/robot/arduino-due/tests_periph_gpio/xunit.xml)

File: dist/tools/compile_and_test_for_board/compile_and_test_for_board.py (hunk 1 above)
Comment (miri64) [reply dist/tools/compile_and_test_for_board/compile_and_test_for_board.py:337]: Done. See 6853d3f7122de764c6deb34c89996ee314e56b94

**Comment by miri64** [comment]
Its now: if there is an error during the test phase: its a failure, otherwise it is an error. Skipping was marked already, passing is the default.

**Review by MrKevinWeiss — APPROVED** [review]
Looks good, ACK, please squash!

**Comment by miri64** [comment]
Squashed.

**Comment by MrKevinWeiss** [comment]
LETS GOOOO!

**Pull request merged**